src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexNodes.cs

```diff
@@ -0,0 +1,1643 @@
+        {
+            switch (index)
+            {
+                case 0: return BackslashToken;
+                case 1: return TypeToken;
+                case 2: return HexText;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// ```\uFFFF``` escape.
+    /// </summary>
+    internal sealed class RegexUnicodeEscapeNode : RegexTypeEscapeNode
+    {
+        public RegexUnicodeEscapeNode(RegexToken backslashToken, RegexToken typeToken, RegexToken hexText)
+            : base(RegexKind.UnicodeEscape, backslashToken, typeToken)
+        {
+            HexText = hexText;
+        }
+
+        internal override int ChildCount => 3;
+
+        public RegexToken HexText { get; }
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+                case 0: return BackslashToken;
+                case 1: return TypeToken;
+                case 2: return HexText;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// ```\'name'``` or ```\&lt;name&gt;``` escape.
+    /// </summary>
+    internal sealed class RegexCaptureEscapeNode : RegexEscapeNode
+    {
+        public RegexCaptureEscapeNode(
+            RegexToken backslashToken, RegexToken openToken, RegexToken captureToken, RegexToken closeToken)
+            : base(RegexKind.CaptureEscape, backslashToken)
+        {
+            OpenToken = openToken;
+            CaptureToken = captureToken;
+            CloseToken = closeToken;
+        }
+
+        internal override int ChildCount => 4;
+
+        public RegexToken OpenToken { get; }
+        public RegexToken CaptureToken { get; }
+        public RegexToken CloseToken { get; }
+
+        internal override RegexNodeOrToken ChildAt(int index)
+        {
+            switch (index)
+            {
+                case 0: return BackslashToken;
+                case 1: return OpenToken;
+                case 2: return CaptureToken;
+                case 3: return CloseToken;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        public override void Accept(IRegexNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    /// <summary>
+    /// ```\k'name'``` or ```\k&lt;name&gt;``` escape.
+    /// </summary>
+    internal sealed class RegexKCaptureEscapeNode : RegexTypeEscapeNode
+    {
+        public RegexKCaptureEscapeNode(
+            RegexToken backslashToken, RegexToken typeToken,
+            RegexToken openToken, RegexToken captureToken, RegexToken closeToken)
+            : base(RegexKind.KCaptureEscape, backslashToken, typeToken)
+        {
+            OpenToken = openToken;
```

Comment:
Could assert on open and close tokens?

Reply:
Consider asserting on open and close tokens.

---
In reply to: [178450369](https://github.com/dotnet/roslyn/pull/23984#discussion_r178450369) [](ancestors = 178450369)